libpromises/policy.c

```diff
@@ -1425,6 +1430,17 @@ Constraint *PromiseAppendConstraint(Promise *pp, const char *lval, Rval rval, bo
         Constraint *old_cp = SeqAt(pp->conlist, i);
         if (strcmp(old_cp->lval, lval) == 0)
         {
+            if (strcmp(old_cp->lval, "ifvarclass") == 0 ||
+                strcmp(old_cp->lval, "if") == 0)
+            {
+                Buffer *grow = BufferNew();
+                BufferAppendF(grow, "(%s).(", old_cp->rval);
+                BufferAppend(grow, RvalScalarValue(rval), strlen(RvalScalarValue(rval)));
+                BufferAppendChar(grow, ')');
+                rval.item = xstrdup(BufferData(grow)); // TODO: will this leak?
```

Comment:
Right, OK.  Besides destroying `cp->rval` we also need to make a new rval, see my new commit which I believe does the right thing.